### PR TITLE
adding tests for transactions v3

### PR DESCRIPTION
**Pull request by Gerson2102**
Adding some tests for txv3, some of them are not working.

# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Bugfix
- Feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no API changes)
- Build-related changes
- Documentation content changes
- Testing
- Other (please describe):

## What is the current behavior?

Resolves: #1574 

 NO

## Other information

The tests that are supposed to fail are working properly. But those tests that have to success, are failing. Maybe I'm doing something wrong here, i dont think that is a coincidence.

### CODE REVIEW

**Comment by Gerson2102** [comment]
@tdelabro I just updated the PR sir.

**Comment by tdelabro** [comment]
@Gerson2102 the CI is failing

File: crates/pallets/starknet/src/tests/invoke_tx.rs [hunk 1]
```diff
@@ -904,3 +907,196 @@ fn given_hardcoded_contract_set_erc20_balance_to_zero() {
...
+        if let starknet_api::transaction::InvokeTransaction::V3(tx) = &mut transaction.tx {
+            tx.signature = sign_message_hash(transaction.tx_hash);
+        };
+
+        assert_ok!(Starknet::invoke(none_origin, transaction));
+    });
+}
+
+#[test]
+fn given_hardcoded_contract_run_invoke_tx_v3_on_openzeppelin_account_then_it_works() {
```
Comment (tdelabro) [inline crates/pallets/starknet/src/tests/invoke_tx.rs:949]: I don't think you have to write test for all the different wallets.
The wallets tests don't change. So everything is already tested by the other tests

Comment (tdelabro) [reply crates/pallets/starknet/src/tests/invoke_tx.rs:949]: What we want you to test is what is constitutive of the txV3, which is the new fee mechanism using STRK

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/pallets/starknet/src/tests/invoke_tx.rs (hunk 1 above)
Comment (Gerson2102) [reply crates/pallets/starknet/src/tests/invoke_tx.rs:949]: Ok, i get it. I'm going to erase the tests related to wallets. Where can i find that new fee mechanism to see if i can understand it and write some tests to them?

**Comment by Gerson2102** [comment]
In the code I'm seeing some tests related to strk and eth fees, i have to do something like that? Or is already tested?

File: crates/pallets/starknet/src/tests/invoke_tx.rs (hunk 1 above)
Comment (tdelabro) [reply crates/pallets/starknet/src/tests/invoke_tx.rs]: > In the code I'm seeing some tests related to strk and eth fees, i have to do something like that? Or is already tested?


I don't think so. Just write one test for each txV3, checking it works correctly

Comment (Gerson2102) [reply crates/pallets/starknet/src/tests/invoke_tx.rs]: > > In the code I'm seeing some tests related to strk and eth fees, i have to do something like that? Or is already tested?
> 
> I don't think so. Just write one test for each txV3, checking it works correctly

I know that i have to do the tests in `invoke_ts.rs` file. But i dont get exactly where is the functionality that i have to test, to see if i can understand it and do something.

Comment (Gerson2102) [reply crates/pallets/starknet/src/tests/invoke_tx.rs]: What do you mean by saying "one test for each txV3"? Like in the `mod.rs` file?

**Comment by tdelabro** [comment]
@Gerson2102 again there is commits that are not your is this PR.

You should rebase your comment onto the head on the main branch. I don't know how you proceed but it make unreviewable, as I'm reviewing changes from other already merged PRs at the same time.

I recommend you give a look to this: https://docs.gitlab.com/ee/topics/git/git_rebase.html

File: crates/pallets/starknet/src/tests/invoke_tx.rs (hunk 1 above)
Comment (tdelabro) [reply crates/pallets/starknet/src/tests/invoke_tx.rs]: invokeV3, DeclareV3, deploy_accountV3
Those are the 3 tx types

**Comment by tdelabro** [comment]
the CI is failing

**Comment by Gerson2102** [comment]
I'm working in this test:

````

#[test]
fn declare_transaction_v3_then_it_works() {
    new_test_ext::<MockRuntime>().execute_with(|| {
        basic_test_setup(2);

        let chain_id = Starknet::chain_id();
        let none_origin = RuntimeOrigin::none();

        let contract_class = get_contract_class("HelloStarknet.casm.json", 1);
        let contract_class_hash = ClassHash(
            StarkFelt::try_from("0x7d2bcb1df4970245665a19b23a4d3877eb86a661e8d98b89afc4531134b99f6").unwrap(),
        );
        let contract_compiled_class_hash = CompiledClassHash(
            StarkFelt::try_from("0x1c02b663e928ed213d3a0fa206efb59182fa2ba41f5c204daa56c4a434b53e5").unwrap(),
        );

        let mut declare_tx = starknet_api::transaction::DeclareTransactionV3 {
            resource_bounds: create_resource_bounds(),
            tip: starknet_api::transaction::Tip::default(),
            signature: TransactionSignature::default(),
            nonce: Nonce(StarkFelt::ZERO),
            class_hash: contract_class_hash,
            compiled_class_hash: contract_compiled_class_hash,
            sender_address: get_account_address(None, AccountType::V1(AccountTypeV1Inner::NoValidate)),
            nonce_data_availability_mode: DataAvailabilityMode::L1,
            fee_data_availability_mode: DataAvailabilityMode::L1,
            paymaster_data: starknet_api::transaction::PaymasterData(vec![StarkFelt::ZERO]),
            account_deployment_data: starknet_api::transaction::AccountDeploymentData(vec![StarkFelt::ZERO]),
        };

        let tx_hash = declare_tx.compute_hash(chain_id, false);
        declare_tx.signature = sign_message_hash(tx_hash);

        let transaction = DeclareTransaction::new(
            starknet_api::transaction::DeclareTransaction::V3(declare_tx),
            tx_hash,
            ClassInfo::new(&contract_class, 1, 1).unwrap(),
        )
        .unwrap();

        assert_ok!(Starknet::declare(none_origin, transaction));
    })
}
````

I'm doing that test based in other test. I just want to declare the tx v3, and check if its working, but I'm getting an error in the assert_ok!(). I dont know if I'm doing something wrong here.


The test I based it on:

````

#[test]
#[ignore]
fn storage_changes_should_revert_on_transaction_revert() {
    new_test_ext::<MockRuntime>().execute_with(|| {
        basic_test_setup(2);

        let chain_id = Starknet::chain_id();
        let none_origin = RuntimeOrigin::none();

        let account_addr = get_account_address(None, AccountType::V1(AccountTypeV1Inner::NoValidate));

        let transaction_revert_class = get_contract_class("TransactionRevert.casm.json", 1);
        let transaction_revert_class_hash = ClassHash(
            StarkFelt::try_from("0x7d2bcb1df4970245665a19b23a4d3877eb86a661e8d98b89afc4531134b99f6").unwrap(),
        );
        let transaction_revert_compiled_class_hash = CompiledClassHash(
            StarkFelt::try_from("0x1c02b663e928ed213d3a0fa206efb59182fa2ba41f5c204daa56c4a434b53e5").unwrap(),
        );

        let mut declare_tx = starknet_api::transaction::DeclareTransactionV2 {
            sender_address: account_addr,
            class_hash: transaction_revert_class_hash,
            compiled_class_hash: transaction_revert_compiled_class_hash,
            nonce: Nonce::default(),
            max_fee: Fee(u128::MAX),
            signature: TransactionSignature(vec![]),
        };

        let tx_hash = declare_tx.compute_hash(chain_id, false);
        declare_tx.signature = sign_message_hash(tx_hash);

        let transaction = DeclareTransaction::new(
            starknet_api::transaction::DeclareTransaction::V2(declare_tx),
            tx_hash,
            ClassInfo::new(&transaction_revert_class, 1, 1).unwrap(),
        )
        .unwrap();

        assert_ok!(Starknet::declare(none_origin, transaction));
        assert_eq!(
            Starknet::contract_class_by_class_hash(transaction_revert_class_hash.0).unwrap(),
            transaction_revert_class
        );

        let salt = ContractAddressSalt(StarkFelt::ZERO);

        let mut invoke_tx = starknet_api::transaction::InvokeTransactionV1 {
            sender_address: account_addr,
            signature: TransactionSignature(vec![]),
            nonce: Nonce(StarkFelt::ONE),
            calldata: Calldata(Arc::new(vec![
                StarkFelt::ONE,
                StarkFelt::try_from(UDC_ADDRESS).unwrap(),  // udc address
                StarkFelt::try_from(UDC_SELECTOR).unwrap(), // deployContract selector
                StarkFelt::try_from("0x4").unwrap(),        // calldata len
                transaction_revert_class_hash.0,            // contract class hash
                salt.0,                                     // salt
                StarkFelt::ONE,                             // unique
                StarkFelt::ZERO,                            // constructor calldata len
            ])),
            max_fee: Fee(u128::MAX),
        };

        let tx_hash = invoke_tx.compute_hash(chain_id, false);
        invoke_tx.signature = sign_message_hash(tx_hash);
        let transaction = InvokeTransaction {
            tx: starknet_api::transaction::InvokeTransaction::V1(invoke_tx),
            tx_hash,
            only_query: false,
        };
        let contract_address: FieldElement = get_udc_deployed_address(
            Felt252Wrapper::from(salt).into(),
            Felt252Wrapper::from(transaction_revert_class_hash).into(),
            &UdcUniqueness::Unique(UdcUniqueSettings {
                deployer_address: Felt252Wrapper::from(account_addr).into(),
                udc_contract_address: FieldElement::from_hex_be(UDC_ADDRESS).unwrap(),
            }),
            &[],
        );
        let contract_address: ContractAddress = Felt252Wrapper::from(contract_address).into();

        // deploy contract
        assert_ok!(Starknet::invoke(RuntimeOrigin::none(), transaction));

        let increase_balance_function_selector = get_selector_from_name("increase_balance").unwrap();

        // create increase balance transaction
        let increase_balance_tx = starknet_api::transaction::InvokeTransactionV1 {
            sender_address: account_addr,
            signature: TransactionSignature(vec![]),
            nonce: Nonce(StarkFelt::TWO),
            max_fee: Fee(u128::MAX),
            calldata: Calldata(Arc::new(vec![
                StarkFelt::ONE,
                contract_address.0.0,
                Felt252Wrapper::from(increase_balance_function_selector).into(),
                StarkFelt::try_from("0x1").unwrap(),
                StarkFelt::try_from("0xa").unwrap(),
            ])),
        };

        // the transaction reverts and returns Ok
        assert_ok!(Starknet::invoke(RuntimeOrigin::none(), increase_balance_tx.clone().into()));

        // the storage value should be 0 after the transaction reverts
        let get_balance_function_selector = get_selector_from_name("get_balance").unwrap();

        let get_balance_function_selector_entrypoint = Felt252Wrapper::from(get_balance_function_selector).into();

        let default_calldata = Calldata(Default::default());

        let balance_value =
            Starknet::call_contract(contract_address, get_balance_function_selector_entrypoint, default_calldata)
                .unwrap();
        assert_eq!(balance_value, vec![Felt252Wrapper::ZERO])
    })
}
````

**Comment by tdelabro** [comment]
@Gerson2102 
https://github.com/keep-starknet-strange/madara/blob/main/crates/pallets/starknet/src/lib.rs#L582
There is a log in this line giving you explanation about what went wrong.
Try to read it. Youn can also replace it with a `println` it may be more easy to see when running tests

File: CHANGELOG.md [hunk 2]
```diff
@@ -59,6 +59,7 @@
 - feat: Remove generic hasher from block hash computation
 - refacto: git submodules removed
 - test: Add pallet test for transaction declare V0
+- test: Deleting tests for wallets
```
Comment (tdelabro) [inline CHANGELOG.md:62]: you need to move this under the `# Next release` section now

File: crates/pallets/starknet/src/tests/invoke_tx.rs [hunk 3]
```diff
@@ -903,3 +905,166 @@ fn given_hardcoded_contract_set_erc20_balance_to_zero() {
...
+                TransactionSource::InBlock,
+                &crate::Call::invoke { transaction: transaction.into() }
+            ),
+            Err(TransactionValidityError::Invalid(InvalidTransaction::BadProof))
+        );
+    })
+}
+
+#[test]
+fn declare_transaction_v3_then_it_works() {
```
Comment (tdelabro) [inline crates/pallets/starknet/src/tests/invoke_tx.rs:990]: move the declare tx test in `declare_tx.rs` test file

File: crates/pallets/starknet/src/tests/invoke_tx.rs [hunk 4]
```diff
@@ -903,3 +905,166 @@ fn given_hardcoded_contract_set_erc20_balance_to_zero() {
...
+            ClassInfo::new(&contract_class, 1, 1).unwrap(),
+        )
+        .unwrap();
+
+        assert_ok!(Starknet::declare(none_origin, transaction));
+    })
+}
+
+#[test]
+fn deploy_transaction_v3_then_it_works() {
```
Comment (tdelabro) [inline crates/pallets/starknet/src/tests/invoke_tx.rs:1035]: same with deploy, there is a dedicated file

**Comment by Gerson2102** [comment]
In `declare_transaction_v3_then_it_works()` I'm having problems because I'm running out of gas. And in the `deploy_transaction_v3_then_it_works()` test i found that i need to first declare the transactions, and then we have to deploy it. So i added the declare part, and I'm running out of gas again.

This is the code for the deploy test:

````
#[test]
fn deploy_transaction_v3_then_it_works() {
    new_test_ext::<MockRuntime>().execute_with(|| {
        basic_test_setup(2);

        let chain_id = Starknet::chain_id();
        let none_origin = RuntimeOrigin::none();

        // Declarar la clase de contrato primero
        let account_addr = get_account_address(None, AccountType::V1(AccountTypeV1Inner::NoValidate));
        let contract_class = get_contract_class("TransactionRevert.casm.json", 1);
        let contract_class_hash = ClassHash(
            StarkFelt::try_from("0x7d2bcb1df4970245665a19b23a4d3877eb86a661e8d98b89afc4531134b99f6").unwrap(),
        );
        let contract_compiled_class_hash = CompiledClassHash(
            StarkFelt::try_from("0x1c02b663e928ed213d3a0fa206efb59182fa2ba41f5c204daa56c4a434b53e5").unwrap(),
        );

        let mut declare_tx = starknet_api::transaction::DeclareTransactionV3 {
            resource_bounds: create_resource_bounds(),
            tip: starknet_api::transaction::Tip::default(),
            signature: TransactionSignature::default(),
            nonce: Nonce(StarkFelt::ZERO),
            class_hash: contract_class_hash,
            compiled_class_hash: contract_compiled_class_hash,
            sender_address: account_addr,
            nonce_data_availability_mode: DataAvailabilityMode::L1,
            fee_data_availability_mode: DataAvailabilityMode::L1,
            paymaster_data: starknet_api::transaction::PaymasterData(vec![StarkFelt::ZERO]),
            account_deployment_data: starknet_api::transaction::AccountDeploymentData(vec![StarkFelt::ZERO]),
        };

        let tx_hash = declare_tx.compute_hash(chain_id, false);
        declare_tx.signature = sign_message_hash(tx_hash);

        let declare_transaction = DeclareTransaction::new(
            starknet_api::transaction::DeclareTransaction::V3(declare_tx),
            tx_hash,
            ClassInfo::new(&contract_class, 1, 1).unwrap(),
        )
        .unwrap();

        assert_ok!(Starknet::declare(none_origin.clone(), declare_transaction));

        // Ahora desplegar la cuenta
        let mut deploy_tx = starknet_api::transaction::DeployAccountTransactionV3 {
            resource_bounds: create_resource_bounds(),
            tip: starknet_api::transaction::Tip::default(),
            signature: TransactionSignature::default(),
            nonce: Nonce(StarkFelt::ZERO),
            class_hash: contract_class_hash,
            contract_address_salt: ContractAddressSalt::default(),
            constructor_calldata: Calldata::default(),
            nonce_data_availability_mode: DataAvailabilityMode::L1,
            fee_data_availability_mode: DataAvailabilityMode::L1,
            paymaster_data: starknet_api::transaction::PaymasterData(vec![StarkFelt::ZERO]),
        };

        let tx_hash = deploy_tx.compute_hash(chain_id, false);
        deploy_tx.signature = sign_message_hash(tx_hash);

        let transaction = DeployAccountTransaction::new(
            starknet_api::transaction::DeployAccountTransaction::V3(deploy_tx),
            tx_hash,
            ContractAddress::default(),
        );

        assert_ok!(Starknet::deploy_account(none_origin, transaction));
    })
}
````

File: crates/pallets/starknet/src/tests/invoke_tx.rs [hunk 5]
```diff
@@ -903,3 +905,166 @@ fn given_hardcoded_contract_set_erc20_balance_to_zero() {
...
+        ));
+
+        let mut transaction = get_invoke_v3_dummy(Starknet::chain_id(), NONCE_ZERO);
+        if let starknet_api::transaction::InvokeTransaction::V3(tx) = &mut transaction.tx {
+            tx.sender_address = contract_address;
+        };
+
+        assert_err!(Starknet::invoke(none_origin, transaction), Error::<MockRuntime>::AccountNotDeployed);
+    })
+}
```
Comment (tdelabro) [inline crates/pallets/starknet/src/tests/invoke_tx.rs:928]: no need for this one

**Review by tdelabro — CHANGES_REQUESTED** [review]

**Comment by tdelabro** [comment]
@Gerson2102 the test you added are failing

**Comment by Gerson2102** [comment]
> @Gerson2102 the test you added are failing

Than you so much for ur guidance so far sir. Still working on it.

**Comment by Gerson2102** [comment]
This is the error that I'm getting on the last test in `declare_tx.rs` file:

````
failures:

---- tests::declare_tx::given_contract_declare_tx3_than_works stdout ----
thread 'tests::declare_tx::given_contract_declare_tx3_than_works' panicked at crates/pallets/starknet/src/tests/declare_tx.rs:63:97:
called `Result::unwrap()` on an `Err` value: ContractClassVersionMismatch { declare_version: TransactionVersion(StarkFelt("0x0000000000000000000000000000000000000000000000000000000000000003")), cairo_version: 0 }


failures:
    tests::declare_tx::given_contract_declare_tx3_than_works
````

**Comment by Gerson2102** [comment]
I dont know which address should I use for the compiled class hash:

````
let compiled_erc20_class_hash = CompiledClassHash(
        StarkFelt::try_from("0x00df4d3042eec107abe704619f13d92bbe01a58029311b7a1886b23dcbb4ea87").unwrap(),
    );
````
I put it different because i was looking at another test that was using different addresses.

File: crates/pallets/starknet/src/tests/declare_tx.rs [hunk 6]
```diff
@@ -9,19 +9,60 @@ use sp_runtime::transaction_validity::{
...
 
+fn create_declare_erc20_v3_transaction(
+    chain_id: Felt252Wrapper,
+    account_type: AccountType,
+    sender_address: Option<ContractAddress>,
+    signature: Option<TransactionSignature>,
+    nonce: Option<Nonce>,
+) -> BlockifierDeclareTransaction {
+    let sender_address = sender_address.unwrap_or_else(|| get_account_address(None, account_type));
+    let erc20_class = get_contract_class("ERC20.json", 0);
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/pallets/starknet/src/tests/declare_tx.rs:34]: you can try declaring `erc20.casm.json`
`    let erc20_class = get_contract_class("erc20.casm.json", 1);`

**Review by tdelabro — CHANGES_REQUESTED** [review]

**Comment by Gerson2102** [comment]
Now im getting an error related to the gas amount:

````
Transaction execution failed: FeeCheckError(MaxL1GasAmountExceeded { max_amount: 10000, actual_amount: 136824 })
thread 'tests::declare_tx::given_contract_declare_tx3_than_works' panicked at crates/pallets/starknet/src/tests/declare_tx.rs:458:9:
Expected Ok(_). Got Err(
    Module(
        ModuleError {
            index: 1,
            error: [
                1,
                0,
                0,
                0,
            ],
            message: Some(
                "TransactionExecutionFailed",
            ),
        },
    ),
)
````
I dont know if i should increment it or something.

**Comment by tdelabro** [comment]
@Gerson2102 yup, you should definitely increment it.

Change the content of `create_resource_bounds`

**Comment by Gerson2102** [comment]
@tdelabro I think I'm done sir. I'm just trying to fix the CI issues.

**Comment by Gerson2102** [comment]
When i run this `npx prettier --write .` is not touching those files that are failing. So, i dont know how to fix it :(

File: CHANGELOG.md [hunk 7]
```diff
@@ -2,6 +2,9 @@
 
 ## Next release
 
+- test: Deleting tests for wallets
+- test: Adding txv3 tests
+
```
Comment (tdelabro) [inline CHANGELOG.md:7]: remove the newline here.

Maybe it will be enough to fix prettier

**Review by tdelabro — CHANGES_REQUESTED** [review]
Mostly it seems you don't have the same version of prettier

**Comment by tdelabro** [comment]
[warn] docs/content/articles/cn/madara-beast-article.md
[warn] docs/contributor-starter-pack.md

I don't know how to fix those either. Prettier is supposed to do it

**Comment by tdelabro** [comment]
@Gerson2102 I will try to pull it myself and run prettier on my computer and then push

**Comment by Gerson2102** [comment]
> @Gerson2102 I will try to pull it myself and run prettier on my computer and then push

No worries, i already fixed it.

File: crates/pallets/starknet/src/tests/invoke_tx.rs [hunk 8]
```diff
@@ -903,3 +905,62 @@ fn given_hardcoded_contract_set_erc20_balance_to_zero() {
         );
     });
 }
+
```
Comment (tdelabro) [inline crates/pallets/starknet/src/tests/invoke_tx.rs:908]: Add a test where invokev3 works

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/pallets/starknet/src/tests/declare_tx.rs [hunk 9]
```diff
@@ -9,19 +9,60 @@ use sp_runtime::transaction_validity::{
...
+        signature: TransactionSignature::default(),
+        nonce: nonce.unwrap_or_default(),
+        class_hash: erc20_class_hash,
+        compiled_class_hash: compiled_erc20_class_hash,
+        sender_address,
+        nonce_data_availability_mode: DataAvailabilityMode::L1,
+        fee_data_availability_mode: DataAvailabilityMode::L1,
+        paymaster_data: starknet_api::transaction::PaymasterData(vec![]),
+        account_deployment_data: starknet_api::transaction::AccountDeploymentData(vec![StarkFelt::ZERO]),
+    });
```
Comment (tdelabro) [inline crates/pallets/starknet/src/tests/declare_tx.rs:54]: https://github.com/keep-starknet-strange/madara/pull/1607#discussion_r1626267632

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 10]
```diff
@@ -2,6 +2,8 @@
 
 ## Next release
 
+- test: Deleting tests for wallets
```
Comment (tdelabro) [inline CHANGELOG.md:5]: you didn't deleted any tests in the end. Did you?

**Review by tdelabro — APPROVED** [review]

File: CHANGELOG.md (hunk 10 above)
Comment (Gerson2102) [reply CHANGELOG.md:5]: I deleted the tests related to the wallets that you told me the other day.

Comment (tdelabro) [reply CHANGELOG.md:5]: after adding them yourself. But in the final PR there is no deletion. So don't include it in the changelog

**Comment by Gerson2102** [comment]
Now all looks green :)

**Comment by tdelabro** [comment]
GG @Gerson2102! You did it! Thanks a lot

**Pull request merged**